core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -551,21 +556,33 @@ class ProducerStateManager(val topicPartition: TopicPartition,
    */
   def oldestSnapshotOffset: Option[Long] = oldestSnapshotFile.map(file => offsetFromFilename(file.getName))
 
+  private def isProducerRetained(producerIdEntry: ProducerIdEntry, logStartOffset: Long): Boolean = {
+    producerIdEntry.lastOffset >= logStartOffset
```

Comment:
@junrao I added some code to do this. See my comment in the commit. I fear we will have to live with the possibility of this causing the LSO to temporarily diverge between replicas. Hopefully this is rare. In any case, there seems little else we can do about it this late in the game (though I'm open to ideas).